libyul/ASTLabelRegistry.cpp

```diff
@@ -0,0 +1,194 @@
+	return m_idToLabelMapping[_id];
+}
+
+std::string_view ASTLabelRegistry::operator[](LabelID const _id) const
+{
+	auto const labelIndex = idToLabelIndex(_id);
+	if (labelIndex == ghostID())
+		return lookupGhost(_id);
+	return m_labels[labelIndex];
+}
+
+std::optional<ASTLabelRegistry::LabelID> ASTLabelRegistry::findIdForLabel(std::string_view const _label) const {
+	if (_label.empty())
+		return emptyID();
+	for (LabelID id = 1; id <= maximumID(); ++id)
+		if ((*this)[id] == _label)
+			return id;
+	return std::nullopt;
+}
+
+std::string_view ASTLabelRegistry::lookupGhost(LabelID const _id) const
+{
+	yulAssert(idToLabelIndex(_id) == ghostID());
+	auto const [it, _] = m_ghostLabelCache.try_emplace(_id, fmt::format("GHOST[{}]", _id));
+	return it->second;
+}
+
+ASTLabelRegistryBuilder::DefinedLabels::DefinedLabels():
+	m_mapping{{"", 0}}
+{}
+
+std::tuple<ASTLabelRegistry::LabelID, bool> ASTLabelRegistryBuilder::DefinedLabels::tryInsert(
+	std::string_view const _label,
+	ASTLabelRegistry::LabelID const _id
+)
+{
+	auto const [it, emplaced] = m_mapping.try_emplace(std::string{_label}, _id);
+	return std::make_tuple(it->second, emplaced);
+}
+
+ASTLabelRegistryBuilder::ASTLabelRegistryBuilder():
+	m_nextID(1)
+{}
+
+ASTLabelRegistryBuilder::ASTLabelRegistryBuilder(ASTLabelRegistry const& _existingRegistry)
+{
+	yulAssert(!_existingRegistry.labels().empty() && _existingRegistry[0].empty());
+	auto const maxId = _existingRegistry.maximumID();
+	for (size_t i = 1; i <= maxId; ++i)
+	{
+		auto const existingLabel = _existingRegistry[i];
+		if (!existingLabel.empty())
+		{
+			if (_existingRegistry.idToLabelIndex(i) == ASTLabelRegistry::ghostID())
+				m_ghosts.push_back(i);
+			else
+			{
+				auto const [_, inserted] = m_definedLabels.tryInsert(_existingRegistry[i], i);
+				yulAssert(inserted);
+			}
+		}
+	}
+	m_nextID = _existingRegistry.maximumID() + 1;
+}
+
+ASTLabelRegistry::LabelID ASTLabelRegistryBuilder::define(std::string_view const _label)
+{
+	yulAssert(!isInGhostFormat(_label));
+	auto const [id, inserted] = m_definedLabels.tryInsert(_label, m_nextID);
+	if (inserted)
+		m_nextID++;
+	return id;
+}
+
+ASTLabelRegistry::LabelID ASTLabelRegistryBuilder::addGhost()
+{
+	m_ghosts.push_back(m_nextID);
+	return m_nextID++;
+}
+
+ASTLabelRegistry ASTLabelRegistryBuilder::build() const
+{
+	auto const& labelToIDMapping = m_definedLabels.labelToIDMapping();
+	yulAssert(labelToIDMapping.contains(""));
+	yulAssert(labelToIDMapping.at("") == 0);
+
+	std::vector<std::string> labels{""};
+	labels.reserve(labelToIDMapping.size());
+	auto const maxLabelId = ranges::max(labelToIDMapping | ranges::views::values);
+	auto const maxGhostId = m_ghosts.empty() ? 0 : m_ghosts.back();
+	std::vector<size_t> idToLabelMapping( std::max(maxLabelId, maxGhostId) + 1, 0);
+	yulAssert(!idToLabelMapping.empty(), "Mapping must at least contain empty label");
+	for (auto const& [label, id]: labelToIDMapping)
+	{
+		// skip empty and ghost
+		if (id < 1)
```

Comment:
Outdated comment. But it will also be self-explanatory without a comment if we use the helper:

```suggestion
		if (id == emptyID())
```